packages/contracts/test/diamond/DiamondTestSetup.sol

```diff
@@ -0,0 +1,186 @@
+
+    // contract types of facets to be deployed
+    Diamond diamond;
+    DiamondCutFacet dCutFacet;
+    DiamondLoupeFacet dLoupe;
+    OwnershipFacet ownerF;
+
+    // interfaces with Facet ABI connected to diamond address
+    IDiamondLoupe ILoupe;
+    IDiamondCut ICut;
+
+    string[] facetNames;
+    address[] facetAddressList;
+
+    address owner;
+    address user1;
+    address contract1;
+
+    // deploys diamond and connects facets
+    function setUp() public virtual {
+
+        owner = generateAddress("Owner", false, 10 ether);
+        user1 = generateAddress("User1", false, 10 ether);
+        contract1 = generateAddress("Contract1", true, 10 ether);
+
+        //deploy facets
+        dCutFacet = new DiamondCutFacet();
+        dLoupe = new DiamondLoupeFacet();
+        ownerF = new OwnershipFacet();
+        facetNames = ["DiamondCutFacet", "DiamondLoupeFacet", "OwnershipFacet"];
+
+        // deploy diamond
+        diamond = new Diamond(owner, address(dCutFacet));
+
+        //upgrade diamond with facets
+
+        //build cut struct
+        FacetCut[] memory cut = new FacetCut[](2);
+
+        cut[0] = FacetCut({
+            facetAddress: address(dLoupe),
+            action: FacetCutAction.Add,
+            functionSelectors: generateSelectors("DiamondLoupeFacet")
+        });
+
+        cut[1] = FacetCut({
+            facetAddress: address(ownerF),
+            action: FacetCutAction.Add,
+            functionSelectors: generateSelectors("OwnershipFacet")
+        });
+
+
+        // initialise interfaces
+        ILoupe = IDiamondLoupe(address(diamond));
+        ICut = IDiamondCut(address(diamond));
+
+        //upgrade diamond
+        vm.startPrank(owner);
+        ICut.diamondCut(cut, address(0x0), "");
+		vm.stopPrank();
+
+        // get all addresses
+        facetAddressList = ILoupe.facetAddresses();
+    }
+}
+
+// tests proper upgrade of diamond when adding a facet
+abstract contract AddManagerFacetSetup is DiamondSetup {
+
+    ManagerFacet managerFacet;
+
+    function setUp() public virtual override {
+        super.setUp();
+        //deploy ManagerFacet
+        managerFacet = new ManagerFacet(owner);
+
+        // bytes4[] memory fromGenSelectors  = generateSelectors("ManagerFacet"); // all length is 51
+        bytes4[] memory selectorsInManagerFacet = new bytes4[](6);
+        selectorsInManagerFacet[0] = getSelector(
+            "setTwapOracleAddress(address)"
+        );
+        selectorsInManagerFacet[1] = getSelector(
+            "setuARTokenAddress(address)"
+        );
+        selectorsInManagerFacet[2] = getSelector(
+            "setDebtCouponAddress(address)"
+        );
+        selectorsInManagerFacet[3] = getSelector(
+            "setIncentiveToUAD(address)"
+        );
+        selectorsInManagerFacet[4] = getSelector(
+            "getExcessDollarsDistributor(address)"
+        );
+        selectorsInManagerFacet[5] = getSelector(
+            "initialize(address)"
+        );
```

Comment:
I originally intended to show that It's possible to use already existing UbiquityAlgorithmicDollarManager contract on Diamond. But when adding function selectors on UADManager to Diamond, it generated one error due to trying to add one function that already exists in Diamond, ex: supportsInterface(bytes4)
So simply wanted to avoid checking duplication of function selectors by selecting some function selectors to add manually

Reply:
> so my opinion for getting selectors with the use of generateSelectors is that if we have some limitations (I am not sure but i guess you were trying to use manual way bc of that), lets just update them as the followings.

As long as there is no duplicate function in a new facet, no need to add manually